pxtblocks/fields/field_utils.ts

```diff
@@ -238,28 +238,46 @@ export function needsTilemapUpgrade(ws: Blockly.Workspace) {
     return !!allTiles.length;
 }
 
-export function upgradeTilemapsInWorkspace(ws: Blockly.Workspace, proj: pxt.TilemapProject) {
-    const allTiles = ws.getVariablesOfType(pxt.sprite.BLOCKLY_TILESET_TYPE).map(model => pxt.sprite.legacy.blocklyVariableToTile(model.name));
-    if (!allTiles.length) return;
-
-    try {
-        Blockly.Events.disable();
-        let customMapping: pxt.Tile[] = [];
-
-        for (const tile of allTiles) {
-            if (tile.qualifiedName) {
-                customMapping[tile.projectId] = proj.resolveTile(tile.qualifiedName);
+export function updateTilemapXml(dom: Element, proj: pxt.TilemapProject) {
+    let needsUpgrade = false;
+
+    const upgradedTileMapping: pxt.Map<pxt.Tile> = {};
+
+    for (const element of dom.children) {
```

Comment:
Just curious if iterating over everything in the dom is the most efficient way to do this.. Would this or would filtering for elements with the variable tag and then iterating through those entries be faster?

Reply:
i don't think it would make a huge difference either way. this is only looping over the top-level elements in the xml file so it should only be the variables and any top-level events like on-start